### ISSUE TITLE
When will  released Forge 1.18.2

### ISSUE THREAD

**Issue by mxdyzmx** [maruohon/litematica#575]
When will it be released Forge 1.18.2?Thanks!

**Comment by eldelassombras** [comment]
> When will it be released Forge 1.18.2?Thanks!

I am interested too

**Comment by maruohon** [comment]
The [CurseForge project page](https://www.curseforge.com/minecraft/mc-mods/litematica-forge) explains my update/port plans. Basically at this time I have no idea when things will be in a usable state, let alone ready for an official release.

**Comment by RFNMC** [comment]
WE NEED IT!!! PLEASE

**Comment by Grey-Wind** [comment]
PLEASE!!!We don't have mods to use shematica files and we can't share our red stone machine easily.

**Comment by aditeya24** [comment]
The forge ports for all of masa's client mods are planned but there are some hold-ups. _Quoted from the discord in #litematica_, "Depending on if I could actually get back to working on the rewrite code at some point and get the 1.12.2 version in a bit better shape for early test ports to 1.19.2 Fabric, then there might also be early Forge test ports soon after that. The main hold-up is that there are a few bigger changes I want to make in 1.12.2 before making any test ports to the later versions. One of those things being the renderer rewrite, which is badly needed on the recent MC versions. The current renderer is a horrible cripled-patch-work-crash-workaround-mess, it's a miracle it works at all anymore", from this you can hopefully understand that it is not that masa doesn't want to make it, there are just somethings that need to be done before the ports are made. Masa also has a full-time job and because of that, he has limited time for modding but he is doing what he can.

> PLEASE!!!We don't have mods to use shematica files and we can't share our red stone machine easily.

Regarding the sharing redstone machines part, it would be recommended that you use fabric if you can to use these tools and some others made to assist in creating redstone contraptions, but if that is not possible, we humbly request that you wait. It is not the matter _if_ the ports are made, it is just a matter of _when_.

**Comment by Grey-Wind** [comment]
> Regarding the sharing redstone machines part, it would be recommended that you use fabric if you can to use these tools and some others made to assist in creating redstone contraptions, but if that is not possible, we humbly request that you wait. It is not the matter if the ports are made, it is just a matter of when.

ok, May I ask when masa can make it? Some mods, like create and TIC, are only available in Forge, so the Forge version of Litematica is useful.

**Comment by aditeya24** [comment]
>May I ask when masa can make it? 

Well, that's the problem, there is no ETA because the progress made on the rewrites are not linear. If any ETA is given, chances are, it will probably not be ready by then. We can only hope it will be ready soon.

**Comment by Grey-Wind** [comment]
> > May I ask when masa can make it?
> 
> Well, that's the problem, there is no ETA because the progress made on the rewrites are not linear. If any ETA is given, chances are, it will probably not be ready by then. We can only hope it will be ready soon.

I wish this could be done sooner

**Comment by kosma** [comment]
We all wish it could be done sooner so questions like these will stop. Masa is just one man with not enough time to do all the requests. Please be patient.

**Comment by Fnige** [comment]
> ok, May I ask when masa can make it? Some mods, like create and TIC, are only available in Forge, so the Forge version of Litematica is useful.

https://modrinth.com/mod/create-fabric and https://modrinth.com/mod/hephaestus

**Comment by ZacSharp** [comment]
<details><summary>Don't take this as a promise for anything. The server for which I need Litematica is on 1.17.1 so there will be a drop in motivation once this is working.</summary> 

![2023-04-03_01 17 21](https://user-images.githubusercontent.com/68165024/229384647-376c4f8c-d882-47c3-b96e-6f1f6d555f5b.png)
</details>

**Comment by maruohon** [comment]
I hope I don't end up having to support multiple totally different 3rd party Forge ports soon... There are already recently created 1.19.2 Forge ports of most of my client mods out there already.

Personally I'm trying to work towards the first "early test port" of the 1.12.2 development code to 1.18/1.19. Currently I have the new malilib running in 1.18.2 Forge, but for Litematica there are still a few util classes and systems I want to clean up in 1.12.2 before I do the first "test port". And that malilib is quite minimally tested as far as the rendering code and some util classes go, as no other mods have been ported yet that use that code.

I call these "early test ports" because they will be one-time-use test branches and will not remain the final branch for eventual releases. I'll just keep cherry-picking the "port commits" from these branches on top of the developing 1.12.2 code as time goes, to make the next test port version at some point. One of the reasons for these "test ports" is to check for stuff that needs to change between MC versions, to see what I can better abstract behind malilib or rewrite in a way that results in a smaller code diff between MC versions and mod loaders. Thus some changes will be incorporated back to 1.12.2, and then later on the next test port will contain fewer porting related changes.

**Comment by ZacSharp** [comment]
Sorry, I didn't see them linked anywhere here on this repo so I assumed the situation was still the same. Can you maybe give links to them so I can avoid duplicate work?
(Will still finish the 1.17.1 port though, it's mostly working by now and it's the version I actually need)

Basically my plan was to go up to 1.19 if mojmap helps enough (no more cross mapping merges, for 1.17 I've avoided that by updating first and now merging) and afterwards maybe have a look at the current 1.12.2 code.

I do these ports mostly for myself, please don't steel your time trying to support my personal-use mess. If you think "useable but buggy" is worse than "nothing at all" just let me know and I'll keep it private (wouldn't be happy about it though, I think there's people who prefer the former).

**Comment by maruohon** [comment]
There are *some kind of* 1.19.2 Forge ports of everything except Item Scroller here: https://github.com/DreamRealmCraft

I haven't tested them at all to see how well they work, but I wouldn't expect any major issues if they are basically the Fabric code slapped on top of Forge.

I don't really want to restrict what people do with the various ports and forks, they are open source after all. It's just that usually when people have issues with whatever version of the mod they are using, they will come to me first for support. That's especially the case with the 1.16.5 Forge ports, since someone else created CurseForge projects for those at one point. I since asked to get ownership of those projects and that person did transfer them to me, so now they even kinda appear to be my ports, even though the text on the CF project page explains the current situation with them. But how many people would do something like *read a bunch of text* :eyes:

So basically those 1.16.5 versions have that config bug that any hotkey changes won't take effect without a re-log or dimension change. And it also affects a small number of other "special load" configs. So dealing with "paste doesn't do anything" type of support questions now involves a question "Are you using the 1.16.5 Forge ports? Those have this bug blahblah...".

And coming back to actually making these ports. The 1.18.2 Forge test port of malilib I did recently I used the Loom Gradle tasks of remapping the code from Yarn to Mojmap names. After that you just need to change the main mod class, the mod loader info file, and manually rename/update all mixin classes, as the Gradle task won't remap mixins. But those 1.19.2 Forge ports I linked have been made using Architectury Loom. So next I also want to look into that more closely as I'm currently not really familiar with it and what all it can do. Because ideally I would like to keep the mods using Yarn names, if it's possible to do that even on the Forge projects via Architectury.

So basically your 1.17.1 port should really be made from the latest 1.17.1 Fabric version mod code for the minimum amount of porting work required. Then you don't have to deal with updating anything like the rendering code manually compared to 1.16.5.

**Comment by ZacSharp** [comment]
> I haven't tested them at all to see how well they work, but I wouldn't expect any major issues if they are basically the Fabric code slapped on top of Forge.

There was one issue specifically with Forge on 1.16 ~, but they would have noticed (immediate crash on schematic world creation) so I guess it's gone.~ (immediate crash when a TE is added to the schematic world) and it is still present on 1.17. Didn't check 1.18+.

> So basically those 1.16.5 versions have that config bug that any hotkey changes won't take effect without a re-log or dimension change. And it also affects a small number of other "special load" configs. So dealing with "paste doesn't do anything" type of support questions now involves a question "Are you using the 1.16.5 Forge ports? Those have this bug blahblah...".

~Uh, I'll try to have a look...~ Fixed

> The 1.18.2 Forge test port of malilib I did recently I used the Loom Gradle tasks of remapping the code from Yarn to Mojmap names. After that you just need to change the main mod class, the mod loader info file, and manually rename/update all mixin classes, as the Gradle task won't remap mixins. 

I feel a bit stupid now. I was using `migrateMappings` and `updateMappings` so I can merge two mojmap mapped branches, but never noticed I could just use it to do the complete port. Just redid my partial port and got [this port](https://github.com/ZacSharp/litematica-forge/commit/d27d509877417b900dd9fddbee484ceb5e41ea88) in 27 minutes, including fixes to the already done [malilib port](https://github.com/ZacSharp/malilib-forge/commit/79ca723ccd76078156fdf51e10ea9a8127964ca0) (plus an awful lot of time for some commit post-signing stunt because I don't have my signing key on the machine I use for the ports)

> So next I also want to look into that more closely as I'm currently not really familiar with it and what all it can do.

Definitely do that. I used archloom once so far (apart from it being used in Baritone) and it worked on both loaders out of the box (pure mixin mod though so I don't know how to deal with modloader specifics).

Those forge ports wiped you from their history so I won't use them. https://github.com/AwangLL/litematica-forge seems to be done properly.

**Comment by maruohon** [comment]
> Those forge ports wiped you from their history so I won't use them.

Yes that was a bit weird. The malilib repo was the only one that kept the original history when I looked.
